diff --git a/CRM/Core/Permission/Joomla.php b/CRM/Core/Permission/Joomla.php
index 090d30ab36e..17a768770bf 100644
--- a/CRM/Core/Permission/Joomla.php
+++ b/CRM/Core/Permission/Joomla.php
@@ -181,8 +181,9 @@ private function getUserGroupPermsAssociations() {
 
     $db->setQuery($query);
 
-    // Load the result as a stdClass object, decoding JSON on the way
-    return json_decode($db->loadObject()->rules);
+    // Joomla gotcha: loadObject returns NULL in the case of no matches.
+    $result = $db->loadObject();
+    return $result ? json_decode($result->rules) : (object) array();
   }
 
   /**
